sdks/python/apache_beam/coders/union_coder.py

```diff
@@ -0,0 +1,123 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pytype: skip-file
+
+import struct
+
+from typing import Iterable
+from typing import List
+from typing import Tuple
+
+from apache_beam.coders import Coder
+from apache_beam.coders.coders import FastCoder
+from apache_beam.typehints import typehints
+
+
+class UnionCoder(FastCoder):
+  def __init__(self, components):
+    # type: (Iterable[Coder]) -> None
+    if not components or not isinstance(components, list):
+      raise ValueError('A valid list of Coders must be provided.')
+
+    if len(components) > 255 or len(components) <= 1:
+      raise ValueError(
+          'The number of components for UnionCoder must be between 2 and 255.')
+
+    self._coders = components
+    self._coder_typehints = {}
+    for i, c in enumerate(self._coders):
+      type_hint = c.to_type_hint()
+      if type_hint in self._coder_typehints:
+        raise ValueError(
+            f'One coder {self._coder_typehints.get(type_hint)} already '
+            f'exists for this type {type_hint}.')
+      else:
+        self._coder_typehints[type_hint] = (struct.pack("B", i), c)
+
+  def _get_coder(self, value) -> Tuple[bytes, Coder]:
+    # have to linearly scan the typehints since type could be composite
+    # simple type(value) does not work
+    typehint_type = None
+    for th in self._coder_typehints:
+      try:
+        typehints.check_constraint(th, value)
+        typehint_type = th
+        break
+      except (typehints.CompositeTypeHintError, typehints.SimpleTypeHintError):
+        pass
+      except Exception as e:
+        raise ValueError(
+            f"cannot check the value {value} type and the error is {e}")
+
+    if typehint_type in self._coder_typehints:
+      return self._coder_typehints.get(typehint_type)
```

Comment:
From the linting precommit - `apache_beam/coders/union_coder.py:68: error: Incompatible return value type (got "Optional[Tuple[bytes, Any]]", expected "Tuple[bytes, Coder]")  [return-value]`